File: app/frontend/src/geometry.ts
export interface Geometry {
  points: [number, number][];
}

/** Migrate legacy {x,y,w,h} rect to 4-point polygon. */
export function normalizeGeometry(g: {
  points?: [number, number][];
  x?: number;
  y?: number;
  w?: number;
  h?: number;
}): Geometry {
  if (g.points && g.points.length >= 3) {
    return { points: g.points.map((p) => [p[0], p[1]] as [number, number]) };
  }
  if (g.x != null && g.y != null && g.w != null && g.h != null) {
    const { x, y, w, h } = g;
    return {
      points: [
        [x, y],
        [x + w, y],
        [x + w, y + h],
        [x, y + h],
      ],
    };
  }
  return { points: [] };
}

export function rectFromDrag(
  x0: number,
  y0: number,
  x1: number,
  y1: number,
  scale: number,
): Geometry {
  const x = Math.round(Math.min(x0, x1) / scale);
  const y = Math.round(Math.min(y0, y1) / scale);
  const w = Math.round(Math.abs(x1 - x0) / scale);
  const h = Math.round(Math.abs(y1 - y0) / scale);
  return normalizeGeometry({ x, y, w, h });
}

export function dist(a: [number, number], b: [number, number]): number {
  return Math.hypot(a[0] - b[0], a[1] - b[1]);
}

File: app/station/core/geometry.py
"""ROI geometry helpers: polygon points with legacy rectangle migration."""

from __future__ import annotations

from typing import Dict, List, Sequence, Tuple

Point = Tuple[int, int]


def rect_to_points(x: int, y: int, w: int, h: int) -> List[Point]:
    """Convert axis-aligned rectangle to 4-point polygon (clockwise)."""
    return [(x, y), (x + w, y), (x + w, y + h), (x, y + h)]


def normalize_geometry_dict(data: Dict) -> Dict:
    """Return ``{"points": [[x,y], ...]}``; migrate legacy ``{x,y,w,h}``."""
    if not isinstance(data, dict):
        raise ValueError(f"geometry must be a dict, got {type(data)}")
    if "points" in data and data["points"]:
        pts = [[int(p[0]), int(p[1])] for p in data["points"]]
        if len(pts) < 3:
            raise ValueError("polygon needs at least 3 points")
        return {"points": pts}
    if all(k in data for k in ("x", "y", "w", "h")):
        x, y, w, h = int(data["x"]), int(data["y"]), int(data["w"]), int(data["h"])
        if w <= 0 or h <= 0:
            raise ValueError(f"invalid rectangle geometry: {data}")
        return {"points": [list(p) for p in rect_to_points(x, y, w, h)]}
    raise ValueError(f"unsupported geometry: {data}")


def geometry_points(geometry: Dict) -> List[Point]:
    """Extract polygon vertices from geometry dict (any supported format)."""
    return [tuple(p) for p in normalize_geometry_dict(geometry)["points"]]


def bbox_from_points(points: Sequence[Point]) -> Tuple[int, int, int, int]:
    """Return (x_min, y_min, x_max, y_max) inclusive of edge pixels."""
    xs = [p[0] for p in points]
    ys = [p[1] for p in points]
    return min(xs), min(ys), max(xs), max(ys)
